### PR TITLE
Together.ai demo app

### PR DESCRIPTION
**Pull request by maxdml**

### CODE REVIEW

File: empyrical_evidences/start_postgres_docker.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  }
+});
+
+if (!process.env.PGPASSWORD) {
+  console.error("Error: PGPASSWORD is not set.");
+  process.exit(1);
+}
+
+try {
+  execSync(`docker run --rm --name=dbos-db --env=POSTGRES_PASSWORD="${process.env.PGPASSWORD}" --env=PGDATA=/var/lib/postgresql/data --volume=/var/lib/postgresql/data -p ${port}:5432 -d sibedge/postgres-plv8`);
```
Comment (qianl15) [inline empyrical_evidences/start_postgres_docker.js:21]: Use ` pgvector/pgvector:pg16` docker image.

File: empyrical_evidences/src/operations.ts [hunk 2]
```diff
@@ -0,0 +1,263 @@
...
+  @Communicator()
+  static async fetchPaper(
+    ctxt: CommunicatorContext,
+    paperURL: string
+  ): Promise<Blob> {
+    const response = await fetch(paperURL);
+    return await response.blob();
+  }
+
+  @Transaction()
```
Comment (qianl15) [inline empyrical_evidences/src/operations.ts:184]: This one should be a communicator.

File: empyrical_evidences/src/operations.ts [hunk 3]
```diff
@@ -0,0 +1,263 @@
...
+  const retriever = vectorStore.asRetriever({
+    filter,
+    callbacks,
+  });
+  return {
+    retriever,
+  };
+}
+
+async function loadVectorStore() {
```
Comment (qianl15) [inline empyrical_evidences/src/operations.ts:245]: Move this to an initializer

File: python/empyrical_evidences/README.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
+<h1 align="center">Empyrical Evidences</h1>
+
+<p align="center">
+  Search hackernews for comment relevant to academic papers.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:4]:
```suggestion
  Search Hacker News for comments relevant to academic papers.
```

File: python/empyrical_evidences/README.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+  Powered by Together.ai and DBOS.
+</p>
+
+## What does the app do
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:8]:
```suggestion
## What It Does
```

File: python/empyrical_evidences/README.md [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+```shell
+export PGPASSWORD=dbos
+python3 start_postgres_docker.py
+```
+
+Then run:
+
+```shell
+python3 -m venv dbos-empyrical-venv
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:112]: Just `.venv`

File: python/empyrical_evidences/README.md [hunk 7]
```diff
@@ -0,0 +1,117 @@
+<h1 align="center">Empyrical Evidences</h1>
+
+<p align="center">
+  Search hackernews for comment relevant to academic papers.
+  Powered by Together.ai and DBOS.
+</p>
+
+## What does the app do
+
+This application uses together.ai inference API to identity relevant topics in academic papers and rank the most relevant related comments from hackernews. It is orchestrated with DBOS Transact and demonstrates how DBOS:
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:10]: What It Does

How It Works

- You upload papers and it embeddings
- You search with multi-agent workflow

What DBOS Does

Instructions + How To Use It

File: python/empyrical_evidences/README.md [hunk 8]
```diff
@@ -0,0 +1,97 @@
...
+
+<p align="center">
+  Search Hacker News for comments relevant to your favorite academic papers.
+  Powered by Together.ai and DBOS.
+</p>
+
+## What It Does
+
+This application uses Together.ai inference APIs to identify relevant topics in academic papers, search Hacker News for related comments and rank them.
+The app is hosted on DBOS Cloud.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:11]:
```suggestion
It is serverlessly hosted on DBOS Cloud.
```

File: python/empyrical_evidences/README.md [hunk 9]
```diff
@@ -0,0 +1,97 @@
...
+
+## How It Works
+
+- Upload a paper. The app uses Together.ai API to query the paper's embeddings and store them in postgres/pgvector.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:15]:
```suggestion
- First, you upload an academic paper. The app uses the Together.ai API to query the paper's embeddings and store them in Postgres using pgvector.
```

File: python/empyrical_evidences/README.md [hunk 10]
```diff
@@ -0,0 +1,97 @@
...
+- Search Hacker News. The app uses a multi-agent workflow and Together.AI inference API. First, it extracts the paper's key topics. Second, it searchs Hacker News for relevant comments. Third, it ranks the comments using Together.ai API.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:16]:
```suggestion
- Then, you search Hacker News for comments relevant to the paper. 
The app does this with a multi-agent workflow that first extracts the paper's key topics, then searches Hacker News for relevant comments, then ranks the comments.
```

File: python/empyrical_evidences/README.md [hunk 11]
```diff
@@ -0,0 +1,97 @@
...
+
+## Why Use DBOS
+
+- Fast Serverless deploy to the cloud.
+- Durable execution: if the application crashes, workflows will be resumed where they left off and Together.ai APIs will not be queried again for already-executed steps.
+- Automatically capture provenance/lineage data for your workflows.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:22]: Rewrite this to say something about automatic tracing

File: python/empyrical_evidences/README.md [hunk 12]
```diff
@@ -0,0 +1,97 @@
...
+The app is hosted on DBOS Cloud.
+
+## How It Works
+
+- Upload a paper. The app uses Together.ai API to query the paper's embeddings and store them in postgres/pgvector.
+- Search Hacker News. The app uses a multi-agent workflow and Together.AI inference API. First, it extracts the paper's key topics. Second, it searchs Hacker News for relevant comments. Third, it ranks the comments using Together.ai API.
+
+## Why Use DBOS
+
+- Fast Serverless deploy to the cloud.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:20]: Flesh this out a bit

File: python/empyrical_evidences/README.md [hunk 13]
```diff
@@ -0,0 +1,97 @@
...
+
+Go to https://api.together.ai/ and sign-up.
+
+### API Key
+
+An API key will be provided for you when you register. You can always retrieve it in your [settings page](https://api.together.ai/settings/api-keys).
+
+## Deploying To DBOS Cloud
+
+### Creating A DBOS Cloud Account
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:36]: Delete this, deploy auto-registers you now

File: python/empyrical_evidences/README.md (hunk 12 above)
Comment (maxdml) [reply python/empyrical_evidences/README.md:20]: let me know how this reads now

File: python/empyrical_evidences/README.md [hunk 14]
```diff
@@ -0,0 +1,94 @@
+<h1 align="center">Empyrical Evidences</h1>
+
+<p align="center">
+  Search Hacker News for comments relevant to your favorite academic papers.
+  Powered by Together.ai and DBOS.
+</p>
+
+## What It Does
+
+This application uses Together.ai inference APIs to identify relevant topics in academic papers, search Hacker News for related comments and rank them.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:10]:
```suggestion
This application uses the Together.ai inference APIs to identify key topics in academic papers, then search Hacker News for related comments and rank them.
```

File: python/empyrical_evidences/README.md [hunk 15]
```diff
@@ -0,0 +1,94 @@
...
+It is serverlessly hosted on DBOS Cloud.
+
+## How It Works
+
+- First, you upload an academic paper. The app uses the Together.ai API to query the paper's embeddings and store them in Postgres using pgvector.
+- Search Hacker News. The app uses a multi-agent workflow and Together.AI inference API. First, it extracts the paper's key topics. Second, it searchs Hacker News for relevant comments. Third, it ranks the comments using Together.ai API.
+
+## Why Use DBOS
+
+- Durable execution: if the application crashes, workflows will be resumed where they left off and Together.ai APIs will not be queried again for already-executed steps.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:20]: We'll workshop this list when you get to the office.

File: python/empyrical_evidences/README.md [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+## Why Use DBOS
+
+### Serverless
+
+- Go from local to production cloud deployment in 30 seconds with a single command.
+- [15x cheaper](https://www.dbos.dev/blog/dbos-vs-lambda-cost), [25x faster](https://www.dbos.dev/blog/dbos-vs-aws-step-functions-benchmark). 0 devops.
+
+### Durable, transactional execution
+
+- Workflows are always resumed where they left off. Save money and time by not querying APIs again for already-executed steps.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:29]:
```suggestion
- If your app is interrupted, workflows automatically resume from where they left off. Save money and time by not querying APIs again for already-executed steps.
```

File: python/empyrical_evidences/README.md [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+- Workflows steps have exactly-once or at-least-once guarantees.
+
+### Observability
+
+- DBOS Transact applications automatically generate and export OpenTelemetry traces.
+- With DBOS Cloud, access a full observability pipeline exporting and collecting your observability data automatically. Access a dashboard to query your data.
+
+## Setup Instructions
+
+### Creating A Together.ai Account
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:39]:
```suggestion
### Creating a Together.ai Account
```

File: python/empyrical_evidences/README.md [hunk 18]
```diff
@@ -0,0 +1,106 @@
...
+- [15x cheaper](https://www.dbos.dev/blog/dbos-vs-lambda-cost), [25x faster](https://www.dbos.dev/blog/dbos-vs-aws-step-functions-benchmark). 0 devops.
+
+### Durable, transactional execution
+
+- Workflows are always resumed where they left off. Save money and time by not querying APIs again for already-executed steps.
+- Workflows steps have exactly-once or at-least-once guarantees.
+
+### Observability
+
+- DBOS Transact applications automatically generate and export OpenTelemetry traces.
```
Comment (kraftp) [inline python/empyrical_evidences/README.md:34]: DBOS automatically generates and exports OpenTelemetry traces for your application

**Pull request merged**